services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/hakemus/KaivuilmoitusSendValidationTest.kt

```diff
@@ -0,0 +1,610 @@
+    }
+
+    @Test
+    fun `fails when start time is not selected`() {
+        val hakemus = hakemus.copy(startTime = null)
+
+        assertThat(hakemus.validateForSend()).failsWith("startTime")
+    }
+
+    @Test
+    fun `fails when end time is not selected`() {
+        val hakemus = hakemus.copy(endTime = null)
+
+        assertThat(hakemus.validateForSend()).failsWith("endTime")
+    }
+
+    @Test
+    fun `fails when end time is before start time`() {
+        val hakemus =
+            hakemus.copy(
+                startTime = DateFactory.getEndDatetime(),
+                endTime = DateFactory.getStartDatetime(),
+            )
+
+        assertThat(hakemus.validateForSend()).failsWith("endTime")
+    }
+
+    @ParameterizedTest
+    @NullSource
+    @EmptySource
+    fun `fails when areas are null or empty`(areas: List<KaivuilmoitusAlue>?) {
+        val hakemus = hakemus.copy(areas = areas)
+
+        assertThat(hakemus.validateForSend()).failsWith("areas")
+    }
+
+    @Test
+    fun `fails when there is no customer`() {
+        val hakemus = hakemus.copy(customerWithContacts = null)
+
+        assertThat(hakemus.validateForSend()).failsWith("customerWithContacts")
+    }
+
+    @Test
+    fun `fails when there is no contractor`() {
+        val hakemus = hakemus.copy(contractorWithContacts = null)
+
+        assertThat(hakemus.validateForSend()).failsWith("contractorWithContacts")
+    }
+
+    @Test
+    fun `succeeds when there is no property developer`() {
+        val hakemus = hakemus.copy(propertyDeveloperWithContacts = null)
+
+        assertThat(hakemus.validateForSend()).succeeds()
+    }
+
+    @Test
+    fun `succeeds when there is no representative`() {
+        val hakemus = hakemus.copy(representativeWithContacts = null)
+
+        assertThat(hakemus.validateForSend()).succeeds()
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = CustomerType::class, names = ["COMPANY", "ASSOCIATION"])
+    fun `fails when the company customer has no registry key`(tyyppi: CustomerType) {
+        val customer =
+            HakemusyhteystietoFactory.create(tyyppi = tyyppi, ytunnus = null).withYhteyshenkilo()
+        val hakemus = hakemus.copy(customerWithContacts = customer)
+
+        assertThat(hakemus.validateForSend()).failsWith("customerWithContacts.ytunnus")
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = CustomerType::class, names = ["PERSON", "OTHER"])
+    fun `succeeds when the person or other customer has no registry key`(tyyppi: CustomerType) {
+        val customer =
+            HakemusyhteystietoFactory.create(tyyppi = tyyppi, ytunnus = null).withYhteyshenkilo()
+        val hakemus = hakemus.copy(customerWithContacts = customer)
+
+        assertThat(hakemus.validateForSend()).succeeds()
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = CustomerType::class, names = ["COMPANY", "ASSOCIATION"])
+    fun `succeeds when the company contractor has no registry key`(tyyppi: CustomerType) {
```

Comment:
This sounds a bit odd but it is according to the specification so it should be okay.